engineering/standards/frontend.md

```diff
@@ -151,11 +151,15 @@ Note: some of our older code still does not conform to this standard.
 
 ### Component spacing and layout
 
-Components should not set any space around their outer-most visual edges — instead the consuming component should be responsible for layout and spacing.
+To preserve modularity and encapsulation, components should not define their own layout-related behavior — instead the consuming component should be responsible for layout and spacing. Components should not set any space around their outer-most visual edges or define their own z-index.
 
 - **margin**: The component's root element should not set any margin.
-
 - **padding**: If the component's root element has border, it's fine to set padding because the border will serve as the outer-most visual edge. But if there's no border, then there should be no padding.
+- **z-index**:
+    - The component's root element should not set any z-index.
+    - It's fine for child elements _within the component_ to set a z-index, but in such cases the component's root element must establish its own [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context). It's best to use `isolation: isolate;` to establish the stacking context because it clearly communicates intent and works without setting z-index. Once your component has its own stacking context, then you're free to set the z-index values within the component without thinking about anything outside the component. You can use simple values like `1` and `2` within the component because everything is encapsulated.
+    - If you want to render a child component with a specific z-index, then prefer to nest the child component inside a DOM element, setting the z-index on the DOM element (not the component). Alteratively if you absolutely must pass a z-index value _into_ a component, then do so by setting something like `z-index: var(--component-name-z-index, auto);` within the component, and then setting something like `--component-name-z-index: 2;` in the parent.
```

Comment:
I assume you're not suggesting using style props here. I'd like to completely avoid style props in our code base since it adds a dom element automatically.

Reply:
This PR does not suggest or prohibit style props. I'm suggesting that components can only accept z-index values via CSS variables. Once a component accepts a CSS value through a CSS variable we can pass a value to it either (a) through style props, or (b) through CSS defined within an enclosing scope. If you want to completely avoid style props in our code base, I think it would be good to open a separate PR proposing a standard for it. We can have a separate discussion there. If we decide to ban style props, we could still make use of the CSS variable approach here. Notice that in this PR I'm suggesting the CSS variable be named `--component-name-z-index` and not `--z-index`. Namespacing the CSS variable within the component name makes it easier for consuming components to define that CSS variable within their CSS instead of through style props. I do think it's worth having a separate conversation about style props as a pattern (because I don't think we're entirely in agreement), but I think the outcome of that conversation needs to affect the standard I'm proposing in this PR.

Reply:
Sounds good